test/base/src/androidTest/kotlin/io/realm/test/shared/ObjectIdTests.kt

```diff
@@ -0,0 +1,222 @@
+package io.realm.test.shared
+
+import io.realm.ObjectId
+import io.realm.Realm
+import io.realm.RealmConfiguration
+import io.realm.RealmInstant
+import io.realm.RealmResults
+import io.realm.entities.Sample
+import io.realm.internal.platform.runBlocking
+import io.realm.query
+import io.realm.query.Sort
+import io.realm.query.find
+import io.realm.test.assertFailsWithMessage
+import io.realm.test.platform.PlatformUtils
+import io.realm.test.util.use
+import kotlinx.coroutines.delay
+import kotlin.test.AfterTest
+import kotlin.test.BeforeTest
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import kotlin.test.assertTrue
+
+class ObjectIdTests {
+
+    private lateinit var tmpDir: String
+    private lateinit var realm: Realm
+
+    @BeforeTest
+    fun setup() {
+        tmpDir = PlatformUtils.createTempDir()
+        val configuration = RealmConfiguration.Builder(schema = setOf(Sample::class))
+            .directory(tmpDir)
+            .build()
+        realm = Realm.open(configuration)
+    }
+
+    @AfterTest
+    fun tearDown() {
+        if (!realm.isClosed()) {
+            realm.close()
+        }
+        PlatformUtils.deleteTempDir(tmpDir)
+    }
+
+    @Test
+    fun invalid_arguments() {
+        assertFailsWithMessage<IllegalArgumentException>("invalid hexadecimal representation of an ObjectId: []") {
+            ObjectId.from("") // empty string
+        }
+
+        assertFailsWithMessage<IllegalArgumentException>("invalid hexadecimal representation of an ObjectId: [garbage]") {
+            ObjectId.from("garbage") // 12 char needed
+        }
+
+        assertFailsWithMessage<IllegalArgumentException>("invalid hexadecimal representation of an ObjectId: [56X1fc72e0c917e9c4714161]") {
+            ObjectId.from("56X1fc72e0c917e9c4714161") // invalid hex value
+        }
+
+        assertFailsWithMessage<IllegalArgumentException>("byte array size must be 12") {
+            ObjectId.from(byteArrayOf(0xCA.toByte(), 0xFE.toByte(), 0xBA.toByte(), 0xBE.toByte()))
+        }
+    }
+
+    @Test
+    fun boundaries() {
+        roundTrip(ObjectId.from("000000000000000000000000")) { value ->
```

Comment:
I am reluctant with just testing against objects create with the constructor 🤔  

Isn't there some way to verify that the actual input is also represented in the underlying bytes. Especially since you don't have negative tests for `equals`. Maybe a test against the `toString` representation for each of the constructors. Basically all of these tests would pass if you didn't store the input argument in the internal bytes.

Reply:
I guess the `toString` test is actually doing exactly this kind of test, sorry. Probably too difficult to do for the non-hexstring variants, so if we just include some negative `equals` test between the objects created with the variants we should probably be good to go.